specs/experimental/gov-proposal-validator.md

```diff
@@ -0,0 +1,581 @@
+`ProposalTypes`.
+
+```solidity
+/// Schema { approvedProposer: address, proposalType: uint8 }
+bytes32 public immutable ATTESTATION_SCHEMA_UID;
+```
+
+`GOVERNOR`
+
+The address of the Optimism Governor contract.
+
+```solidity
+IOptimismGovernor public immutable GOVERNOR;
+```
+
+`minimumVotingPower`
+
+The minimum voting power a delegate must have in order to be eligible for approving a proposal.
+
+```solidity
+uint256 public minimumVotingPower;
+```
+
+`proposalDistributionThreshold`
+
+The maximum amount of tokens a proposal can request.
+
+```solidity
+uint256 public proposalDistributionThreshold;
+```
+
+`votingCycles`
+
+A mapping that stores the data for each voting cycle.
+
+```solidity
+mapping(uint256 => VotingCycleData) public votingCycles;
+```
+
+`_proposals`
+
+A mapping that stores each submitted proposal's data based on its `proposalHash`. The proposal hash is produced by hashing
+the ABI encoded values of specific proposal params, see [Proposal uniqueness](#proposal-uniqueness).
+
+```solidity
+mapping(bytes32 => ProposalSubmissionData) private _proposals;
+```
+
+`_proposalTypesData`
+
+A mapping that stores data related to each proposal type.
+
+```solidity
+mapping(ProposalType => ProposalTypeData) private _proposalTypesData;
+```
+
+### Structs
+
+`ProposalSubmissionData`
+
+A struct that holds all the data for a single proposal. Consists of:
+
+- `proposer`: The address that submitted the proposal
+- `proposalType`: The type of the proposal
+- `inVoting`: Returns true if the proposal has already been submitted for voting
+- `delegateApprovals`: Mapping of addresses that approved the specific proposal
+- `approvalsCounter`: The number of approvals the specific proposal has received
+- `timestamp`: The timestamp of the proposal submission
+
+```solidity
+struct ProposalSubmissionData {
+    address proposer;
+    ProposalType proposalType;
+    bool inVoting;
+    mapping(address => bool) delegateApprovals;
+    uint256 approvalsCounter;
+    uint256 timestamp;
+}
+```
+
+`ProposalTypeData`
+
+A struct that holds data for each proposal type.
+
+- `requiredApprovals`: The number of approvals each proposal type requires in order to be able to move for voting.
+- `proposalVotingModule`: The voting module (proposal type configurator) that can be used for each proposal type. This
+is set by the owner on initialize.
+
+```solidity
+struct ProposalTypeData {
+    uint256 requiredApprovals;
+    uint8 proposalVotingModule;
+}
+```
+
+`VotingCycleData`
+
+A struct that stores data related to the voting cycle.
+
+- `startingBlock`: The block number/timestamp that the voting cycle starts
+- `duration`: The duration of the specific voting cycle
+- `votingCycleDistributionLimit`: A more general distribution amount limit tied to the voting cycle
+
+```solidity
+struct VotingCycleData {
+    uint256 startingBlock;
+    uint256 duration;
+    uint256 votingCycleDistributionLimit;
```

Comment:
So I guess we could put the `approvalThreshold` here to do it per-cycle?  I guess the issue would then be the OF needs to set it quickly, or we might have the wrong value used in a `canApprove` check.

Maybe we can block delegate approvals until the threshold is set. So the flow would be:
- initial voting cycles have a new `approvalThreshold` field, all are initialized to 0
- once a cycle has started and we know what the threshold should be and set to that value
- `approveProposal` requires the above step was complete/nonzero value

Does that feel doable?  Any other ideas?